### PR TITLE
caddy: update caddy to v2.4.6

### PR DESCRIPTION
**Pull request by kevinwojo**
Upgrade caddy from v2.0.0 to v2.4.6. This will help some customers who are looking to take advantage of Caddy's support forums and take advantage of improvements, fixes, and new features within Caddy.

*I did this in my "free" time*

### Checklist

* [x] Sister [deploy-sourcegraph](https://github.com/sourcegraph/deploy-sourcegraph) change: N/A - caddy not shipped with k8s
* [x] All images have a valid tag and SHA256 sum
### Test plan

Spun up a cloud instance and tested this change. Confirmed LetEncrypt functionality as well.

```
root@docker-compose-dev:~/deploy-sourcegraph-docker/docker-compose# docker ps
CONTAINER ID        IMAGE                                          COMMAND                  CREATED             STATUS                    PORTS                                                                                                                                              NAMES
3d8f06c73644        caddy:alpine                                   "caddy run --config …"   40 seconds ago      Up 38 seconds             0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp, 2019/tcp                                                                                                 caddy
4a70eace735b        sourcegraph/frontend:3.38.0                    "/sbin/tini -- /usr/…"   10 minutes ago      Up 10 minutes (healthy)                                                                                                                                                      sourcegraph-frontend-0
bd31532c987f        sourcegraph/frontend:3.38.0                    "/sbin/tini -- /usr/…"   10 minutes ago      Up 10 minutes (healthy)                                                                                                                                                      sourcegraph-frontend-internal
b94fb337aa40        sourcegraph/search-indexer:3.38.0              "/sbin/tini -- zoekt…"   10 minutes ago      Up 10 minutes                                                                                                                                                                zoekt-indexserver-0
239020ef1a53        sourcegraph/codeinsights-db:3.38.0             "docker-entrypoint.s…"   10 minutes ago      Up 10 minutes             5432/tcp                                                                                                                                           codeinsights-db
b1af89460f8b        sourcegraph/repo-updater:3.38.0                "/sbin/tini -- /usr/…"   10 minutes ago      Up 10 minutes                                                                                                                                                                repo-updater
4fab3a2e470c        sourcegraph/indexed-searcher:3.38.0            "/sbin/tini -- /bin/…"   10 minutes ago      Up 10 minutes (healthy)                                                                                                                                                      zoekt-webserver-0
afe4efd07088        sourcegraph/worker:3.38.0                      "/sbin/tini -- /usr/…"   10 minutes ago      Up 10 minutes             3189/tcp                                                                                                                                           worker
20638534f3d0        sourcegraph/gitserver:3.38.0                   "/sbin/tini -- /usr/…"   10 minutes ago      Up 10 minutes                                                                                                                                                                gitserver-0
fcbf799061ac        sourcegraph/codeintel-db:3.38.0                "/postgres.sh"           12 minutes ago      Up 10 minutes (healthy)   5432/tcp                                                                                                                                           codeintel-db
13d04748f164        sourcegraph/syntax-highlighter:3.38.0          "sh -c '/http-server…"   12 minutes ago      Up 10 minutes (healthy)   9238/tcp                                                                                                                                           syntect-server
9c49bd8c345a        sourcegraph/prometheus:3.38.0                  "/bin/prom-wrapper"      12 minutes ago      Up 10 minutes             0.0.0.0:9090->9090/tcp                                                                                                                             prometheus
9f3e6c838255        sourcegraph/postgres-12-alpine:3.38.0          "/postgres.sh"           12 minutes ago      Up 10 minutes (healthy)   5432/tcp                                                                                                                                           pgsql
b4b29c04a38f        sourcegraph/minio:3.38.0                       "/usr/bin/docker-ent…"   12 minutes ago      Up 10 minutes (healthy)   9000/tcp                                                                                                                                           minio
34747349f479        sourcegraph/redis-store:3.38.0                 "/sbin/tini -- redis…"   12 minutes ago      Up 10 minutes             6379/tcp                                                                                                                                           redis-store
245577d6bed9        sourcegraph/symbols:3.38.0                     "/sbin/tini -- /usr/…"   12 minutes ago      Up 10 minutes (healthy)   3184/tcp                                                                                                                                           symbols-0
668d8f5676bf        sourcegraph/jaeger-all-in-one:3.38.0           "/go/bin/all-in-one-…"   12 minutes ago      Up 10 minutes             5775/udp, 0.0.0.0:5778->5778/tcp, 0.0.0.0:6831-6832->6831-6832/tcp, 0.0.0.0:14250->14250/tcp, 6831-6832/udp, 0.0.0.0:16686->16686/tcp, 14268/tcp   jaeger
db8c0f58feee        sourcegraph/redis-cache:3.38.0                 "/sbin/tini -- redis…"   12 minutes ago      Up 10 minutes             6379/tcp                                                                                                                                           redis-cache
74d727d8f670        sourcegraph/precise-code-intel-worker:3.38.0   "/sbin/tini -- /usr/…"   12 minutes ago      Up 10 minutes (healthy)   3188/tcp                                                                                                                                           precise-code-intel-worker
04f7712089bd        sourcegraph/github-proxy:3.38.0                "/sbin/tini -- /usr/…"   12 minutes ago      Up 10 minutes                                                                                                                                                                github-proxy
a40a4ecdbf1a        sourcegraph/cadvisor:3.38.0                    "/usr/bin/cadvisor -…"   12 minutes ago      Up 10 minutes (healthy)   8080/tcp                                                                                                                                           cadvisor
ca66337c0849        sourcegraph/grafana:3.38.0                     "/entry.sh"              12 minutes ago      Up 10 minutes             3000/tcp, 0.0.0.0:3370->3370/tcp                                                                                                                   grafana
4f98fcc07b43        sourcegraph/postgres_exporter:3.38.0           "/usr/local/bin/post…"   12 minutes ago      Up 10 minutes             9187/tcp                                                                                                                                           codeintel-db-exporter
80058beae376        sourcegraph/postgres_exporter:3.38.0           "/usr/local/bin/post…"   12 minutes ago      Up 10 minutes             9187/tcp                                                                                                                                           pgsql-exporter
46ce41805212        sourcegraph/searcher:3.38.0                    "/sbin/tini -- /usr/…"   12 minutes ago      Up 10 minutes (healthy)                                                                                                                                                      searcher-0
root@docker-compose-dev:~/deploy-sourcegraph-docker/docker-compose# docker exec -it 3d8f06c73644 /bin/sh
/srv # caddy version
v2.4.6 h1:HGkGICFGvyrodcqOOclHKfvJC0qTU7vny/7FhYp9hNw=
```

### CODE REVIEW

**Review by michaellzc — APPROVED** [review]

**Comment by michaellzc** [comment]
should we cherry-pick this to dev just in case?

File: docker-compose/docker-compose.yaml [hunk 1]
```diff
@@ -68,7 +68,7 @@ services:
   # https://caddyserver.com/docs/caddyfile
   caddy:
     container_name: caddy
-    image: 'index.docker.io/caddy/caddy:alpine@sha256:83d9aa7a5f1bbcc0fc1b4720c183a5ec53dae7dc5d9fa555daf3db345010e7f9'
+    image: 'index.docker.io/caddy:alpine@sha256:b5a59725783bab0d65803f87028c68dd6611ca6184040bd98b18797cbe26bdd9'
```
Comment (caugustus-sourcegraph) [inline docker-compose/docker-compose.yaml:71]: Thoughts on including the 2.4.6 version reference in the image so it's a bit more self-documenting?

**Review by caugustus-sourcegraph — COMMENTED** [review]
Looks like dev has been using 2.4.5 so should be good: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/dev/caddy.sh?L9
(edit: That was unrelated to Michael's comment! It's probably worth it to update the dev version as well just for consistency)

Did you test the upgrade path and with all our built-in configurations?

File: docker-compose/docker-compose.yaml (hunk 1 above)
Comment (DaedalusG) [reply docker-compose/docker-compose.yaml:71]: I'd love this -- I always have a hard time backtracking from the sha

**Comment by kevinwojo** [comment]
> Looks like dev has been using 2.4.5 so should be good: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/dev/caddy.sh?L9 (edit: That was unrelated to Michael's comment! It's probably worth it to update the dev version as well just for consistency)
> 
> Did you test the upgrade path and with all our built-in configurations?

@caugustus-sourcegraph: 

Great question! I worked my way through the various CaddyFiles we include in our configuration including the stock plain HTTP configuration. The upgrade path would be 3.38.0 -> 3.39.0. We're using the base features of Caddy2 as a simple reverse proxy. Our built-in configurations that we ship are pretty minimal. Going through the various configurations yielded positive results.

**Default HTTP config**
```
curl -I dsgdct.spicyinfra.com
HTTP/1.1 302 Found
Cache-Control: no-cache, max-age=0
Content-Type: text/html; charset=utf-8
Date: Tue, 29 Mar 2022 10:49:31 GMT
Location: /sign-in?returnTo=%2F
Server: Caddy
Vary: Authorization
Vary: Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Trace: 
X-Xss-Protection: 1; mode=block
```

**LetsEncrypt HTTPS (production)**
```
curl -I https://dsgdct.spicyinfra.com
HTTP/2 302 
cache-control: no-cache, max-age=0
content-type: text/html; charset=utf-8
date: Tue, 29 Mar 2022 10:52:32 GMT
location: /sign-in?returnTo=%2F
server: Caddy
vary: Authorization
vary: Cookie
x-content-type-options: nosniff
x-frame-options: DENY
x-trace: 
x-xss-protection: 1; mode=block
```

**Custom HTTPS**

```
curl -I https://sourcegraph.dev.local
HTTP/2 302 
cache-control: no-cache, max-age=0
content-type: text/html; charset=utf-8
date: Tue, 29 Mar 2022 11:28:43 GMT
location: /sign-in?returnTo=%2F
server: Caddy
vary: Authorization
vary: Cookie
x-content-type-options: nosniff
x-frame-options: DENY
x-trace: 
x-xss-protection: 1; mode=block
```

image

File: pure-docker/deploy-caddy.sh [hunk 2]
```diff
@@ -32,5 +32,5 @@ docker run --detach \
     -p 0.0.0.0:443:443 \
     -v $VOLUME:/caddy-storage \
     --mount type=bind,source="$(pwd)"/../caddy/builtins/http.Caddyfile,target=/etc/caddy/Caddyfile \
-    index.docker.io/caddy/caddy:alpine@sha256:83d9aa7a5f1bbcc0fc1b4720c183a5ec53dae7dc5d9fa555daf3db345010e7f9
+    index.docker.io/caddy:alpine@sha256:b5a59725783bab0d65803f87028c68dd6611ca6184040bd98b18797cbe26bdd9
```
Comment (caugustus-sourcegraph) [inline pure-docker/deploy-caddy.sh:35]:
```suggestion
    index.docker.io/caddy:2.4.6-alpine@sha256:b5a59725783bab0d65803f87028c68dd6611ca6184040bd98b18797cbe26bdd9
```

File: redis/cache-redis.conf [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# allow access from all instances
```
Comment (caugustus-sourcegraph) [inline redis/cache-redis.conf:1]: Looks like the redis conf changes snuck in by accident!

Comment (kevinwojo) [reply redis/cache-redis.conf:1]: !!! NO!!!! bad redis conf!

**Review by caugustus-sourcegraph — APPROVED** [review]
🎉

**Pull request merged**